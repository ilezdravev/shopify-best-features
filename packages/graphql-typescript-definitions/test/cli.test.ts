import {resolve} from 'path';
import {exec} from 'child_process';

const scriptPath = resolve(__dirname, '../bin/graphql-typescript-definitions');
const rootFixturePath = resolve(__dirname, 'fixtures');

describe('cli', () => {
  it('succeeds when there are no fixture errors', async () => {
    expect(await execDetails(cliCommandForFixtureDirectory('all-clear'))).toMatchSnapshot();
  });

  it('fails when there are syntax errors', async () => {
    expect(await execDetails(cliCommandForFixtureDirectory('malformed-query'))).toMatchSnapshot();
    expect(await execDetails(cliCommandForFixtureDirectory('missing-schema'))).toMatchSnapshot();
  });
});

function execDetails(command: string) {
  return new Promise((resolve) => {
    exec(command, (error, stdout, stderr) => {
      resolve({error, stdout, stderr});
    });
  });
}

function cliCommandForFixtureDirectory(fixture: string) {
  const fixtureDirectory = resolve(rootFixturePath, fixture);
  return [
    scriptPath,
    `'${resolve(fixtureDirectory, 'documents/**/*.graphql')}'`,
    `--schema-path '${resolve(fixtureDirectory, 'schema.json')}'`,
  ].join(' ');
}
