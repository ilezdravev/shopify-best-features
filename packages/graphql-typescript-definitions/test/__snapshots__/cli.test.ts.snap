// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cli fails when there are syntax errors 1`] = `
Object {
  "error": Object {
    "cmd": "packages/graphql-typescript-definitions/bin/graphql-typescript-definitions 'packages/graphql-typescript-definitions/test/fixtures/malformed-query/documents/**/*.graphql' --schema-path 'packages/graphql-typescript-definitions/test/fixtures/malformed-query/schema.json'",
    "code": 1,
    "killed": false,
    "signal": null,
  },
  "stderr": "",
  "stdout": " ERROR  Syntax Error: Expected {, found Name \\"name\\"
GraphQLError: Syntax Error: Expected {, found Name \\"name\\"
    at syntaxError (node_modules/graphql/error/syntaxError.js)
    at expect (node_modules/graphql/language/parser.js)
    at many (node_modules/graphql/language/parser.js)
    at parseSelectionSet (node_modules/graphql/language/parser.js)
    at parseFragmentDefinition (node_modules/graphql/language/parser.js)
    at parseExecutableDefinition (node_modules/graphql/language/parser.js)
    at parseDefinition (node_modules/graphql/language/parser.js)
    at parseDocument (node_modules/graphql/language/parser.js)
    at Object.parse (node_modules/graphql/language/parser.js)
    at Builder.<anonymous> (packages/graphql-typescript-definitions/lib/index.js)

",
}
`;

exports[`cli fails when there are syntax errors 2`] = `
Object {
  "error": Object {
    "cmd": "packages/graphql-typescript-definitions/bin/graphql-typescript-definitions 'packages/graphql-typescript-definitions/test/fixtures/missing-schema/documents/**/*.graphql' --schema-path 'packages/graphql-typescript-definitions/test/fixtures/missing-schema/schema.json'",
    "code": 1,
    "killed": false,
    "signal": null,
  },
  "stderr": "",
  "stdout": " ERROR  Error parsing 'packages/graphql-typescript-definitions/test/fixtures/missing-schema/schema.json':

ENOENT: no such file or directory, open 'packages/graphql-typescript-definitions/test/fixtures/missing-schema/schema.json'
Error: Error parsing 'packages/graphql-typescript-definitions/test/fixtures/missing-schema/schema.json':

ENOENT: no such file or directory, open 'packages/graphql-typescript-definitions/test/fixtures/missing-schema/schema.json'
    at Builder.<anonymous> (packages/graphql-typescript-definitions/lib/index.js)
    at Generator.throw (<anonymous>)
    at rejected (packages/graphql-typescript-definitions/lib/index.js)
    at <anonymous>

",
}
`;

exports[`cli fails when there are unused types 1`] = `
Object {
  "error": Object {
    "cmd": "packages/graphql-typescript-definitions/bin/graphql-typescript-definitions 'packages/graphql-typescript-definitions/test/fixtures/missing-types/documents/**/*.graphql' --schema-path 'packages/graphql-typescript-definitions/test/fixtures/missing-types/schema.json'",
    "code": 1,
    "killed": false,
    "signal": null,
  },
  "stderr": "",
  "stdout": "
 ERROR  Error in packages/graphql-typescript-definitions/test/fixtures/missing-types/documents/Query.graphql: an invalid type was provided for key 'role'. This probably means there is a typo in your document or that one of your input types has been renamed.
Error: Error in packages/graphql-typescript-definitions/test/fixtures/missing-types/documents/Query.graphql: an invalid type was provided for key 'role'. This probably means there is a typo in your document or that one of your input types has been renamed.
    at Builder.<anonymous> (packages/graphql-typescript-definitions/lib/index.js)
    at Generator.next (<anonymous>)
    at packages/graphql-typescript-definitions/lib/index.js
    at new Promise (<anonymous>)
    at __awaiter (packages/graphql-typescript-definitions/lib/index.js)
    at Promise.all.Object.keys.map (packages/graphql-typescript-definitions/lib/index.js)
    at Array.map (<anonymous>)
    at Builder.<anonymous> (packages/graphql-typescript-definitions/lib/index.js)
    at Generator.next (<anonymous>)
    at packages/graphql-typescript-definitions/lib/index.js

",
}
`;

exports[`cli succeeds when there are no fixture errors 1`] = `
Object {
  "error": null,
  "stderr": "",
  "stdout": "
 BUILT  packages/graphql-typescript-definitions/test/fixtures/all-clear/documents/Query.graphql → packages/graphql-typescript-definitions/test/fixtures/all-clear/documents/Query.graphql.d.ts
 BUILT  packages/graphql-typescript-definitions/test/fixtures/all-clear/documents/Fragment.graphql → packages/graphql-typescript-definitions/test/fixtures/all-clear/documents/Fragment.graphql.d.ts
",
}
`;
